BRICR-to-v2.0/utils.py

```diff
@@ -0,0 +1,104 @@
+from io import StringIO
+
+from lxml import etree
+
+BUILDINGSYNC_URI = 'http://buildingsync.net/schemas/bedes-auc/2019'
+NAMESPACES = {
+    'auc': BUILDINGSYNC_URI
+}
+
+def children_sorter_factory(schema, tree, element):
+    """returns a function for getting a key value for sorting elements
+
+    Used to sort an elements children after inserting a new element to ensure
+    it meets the schema sequence specification
+
+    :param schema: xmlschema.XmlSchema, the schema to follow
+    :param tree: ElementTree, the tree from which the element came from
+    :param element: Element, the element whose children are to be sorted
+    """
+    # get the element from the schema
+    element_path = tree.getpath(element)
+    schema_element = schema.find(element_path)
+    if schema_element is None:
+        raise Exception(f'Unable to find path in schema: "{element_path}"')
+
+    ordered_children = [child.name for child in schema_element.iterchildren()]
+
+    # construct a function for sorting an element's children by returning the index of the
+    # child according to the ordering of the schema
+    def _getkey(element):
+        if isinstance(element, etree._Comment):
+            # put comments at the end
+            return 100
+        elif not isinstance(element, etree._Element):
+            raise Exception(f'Unknown type while sorting: "{type(element)}"')
+
+        if element.tag not in ordered_children:
+            # a more helpful exception that the one raised by .index()
+            raise Exception(f'Failed to find "{element.tag}" in {ordered_children}')
+
+        return ordered_children.index(element.tag)
+
+    return _getkey
+
+
+def sort_element(schema, tree, element):
+    """Sorts an element's children in place"""
+    getter = children_sorter_factory(schema, tree, element)
+    element[:] = sorted(element, key=getter)
+
+
+def remove_dupes(element):
+    """Removes duplicate children
+    !! assumes that element's children are already in sorted order
+    be careful which elements this is run on as it might find false positives that it removes
+    """
+    prev_child_text = ''
+    prev_child_tag = ''
+    prev_child = None
+    for child in element:
+        if child.text == prev_child_text and child.tag == prev_child_tag:
+            element.remove(child)
+        elif child.tag == prev_child_tag and 'AnnualSavingsSourceEnergy' in child.tag:
+            # keep the one with a smaller value (being conservative)
+            if float(prev_child_text) < float(child.text):
+                element.remove(child)
+            else:
+                element.remove(prev_child)
+
+        prev_child_text = child.text
+        prev_child_tag = child.tag
+        prev_child = child
```

Comment:
@nllong Please take a careful look at this function as it might be pretty dangerous (can delete a lot of data). Looking at the results it seems to be working as intended though.

Reply:
Also, can you confirm that when there are duplicate AnnauSavingsSourceEnergy, it's ok to choose the one with a smaller value?

Reply:
I'll let @corymosiman12 answer the second question... 

The code looks right though.

Reply:
Sorry for the slow reply.  The main reasoning for this is because there were 134 files which had this field as duplicate (though it should not).  The percent difference between the majority of the values (130/134 files) was < 1%, so we assumed it was ok to just keep the 'smallest' savings, as these are most conservative.

![image](https://user-images.githubusercontent.com/28991233/79417170-8607a700-7f6e-11ea-9d08-98e0d65fda1b.png)